arroyo-state/src/parquet.rs

```diff
@@ -479,19 +712,162 @@ impl ParquetBackend {
+                .as_any()
+                .downcast_ref::<arrow_array::UInt64Array>()
+                .unwrap();
+            let time_array = batch
+                .column(1)
+                .as_any()
+                .downcast_ref::<arrow_array::TimestampMicrosecondArray>()
+                .expect("Column 1 is not a TimestampMicrosecondArray");
+            let key_array = batch
+                .column(2)
+                .as_any()
+                .downcast_ref::<arrow_array::BinaryArray>()
+                .unwrap();
+            let value_array = batch
+                .column(3)
+                .as_any()
+                .downcast_ref::<arrow_array::BinaryArray>()
+                .unwrap();
+            let operation_array = batch
+                .column(4)
+                .as_any()
+                .downcast_ref::<arrow_array::UInt32Array>()
+                .unwrap();
+            for index in 0..num_rows {
+                let key_hash = key_hash_array.value(index);
+                if !range.contains(&key_hash) {
+                    continue;
+                }
+
+                let timestamp = from_micros(time_array.value(index) as u64);
+
+                let key = key_array.value(index).to_owned();
+                let value = value_array.value(index).to_owned();
+                let operation = operation_array.value(index).into();
+
+                result.push(BlindTuple {
+                    key_hash,
+                    timestamp,
+                    key,
+                    value,
+                    operation,
+                });
             }
         }
         result
     }
 }
 
-struct ParquetWriter {
+pub struct ParquetCompressedFileWriter {
+    table_char: char,
+    epoch: u32,
+    task_info: TaskInfo,
+    builder: RecordBatchBuilder,
+}
+
+impl ParquetCompressedFileWriter {
+    pub fn new(table_char: char, epoch: u32, task_info: TaskInfo) -> Self {
+        ParquetCompressedFileWriter {
+            table_char,
+            epoch,
+            task_info,
+            builder: RecordBatchBuilder::default(),
+        }
+    }
+
+    pub(crate) fn write(
+        &mut self,
+        key_hash: u64,
+        timestamp: SystemTime,
+        key: Vec<u8>,
+        data: Vec<u8>,
+        operation: DataOperation,
+    ) {
+        self.builder
+            .insert(key_hash, timestamp, key, data, operation);
+    }
+
+    async fn upload_record_batch(
+        key: &str,
+        record_batch: arrow_array::RecordBatch,
+        storage: &StorageProvider,
+    ) -> Result<usize> {
+        let props = WriterProperties::builder()
+            .set_compression(parquet::basic::Compression::ZSTD(ZstdLevel::default()))
+            .set_statistics_enabled(EnabledStatistics::None)
+            .build();
+        let cursor = Vec::new();
+        let mut writer = ArrowWriter::try_new(cursor, record_batch.schema(), Some(props)).unwrap();
+        writer.write(&record_batch).expect("Writing batch");
```

Comment:
Given this returns a Result, I'd just add a `?` rather than the expect and unwrap calls.